### DISCUSSION TITLE
apply_redactions() erases neighboring text which is not redacted

### DISCUSSION THREAD

**Discussion by tigrankh** [pymupdf/PyMuPDF#1810] [General]
Hi,

I'm facing a strange issue when i apply redaction on a PDF.
I have the following text:
![image](https://user-images.githubusercontent.com/9495111/178612119-eba6f51a-95d6-4b92-8673-a018fed8a867.png)

in my experiment i redact the word "von", which overlaps with bbox of the word "Niederschrift" below it and erases some of its letters, as shown below:
![image](https://user-images.githubusercontent.com/9495111/178612773-223bfd20-a2a0-46cb-9118-e06ee0c7bd07.png)

In one of your github issues I saw you recommended using `fitz.TOOLS.set_small_glyph_heights(True)`, which updated the bboxes and they are not overlapping anymore.

However, when i apply redaction to the word "von" it still erases the word below.
I'm suspecting that apply_redactions() isn't just operating on the bbox rectangle, but can't really say for sure. bbox coords of those words do not overlap at all

Could you please help with this ?

**Comment by JorjMcKie** [comment]
There is a solution to this, but let me first move this to the "Discussions" tab.

**Comment by JorjMcKie** [comment]
I assume you have used the lower rectangle produced after `fitz.TOOLS.set_small_glyph_heights(True)` for the redaction as well?

The reason for this effect is the logic behind the redaction algorithm (inside MuPDF code): it is based on characters **overlapping** redaction rectangles - as opposed to characters being **contained**. If the PDF creator was unfortunate enough choosing line distances much smaller than the font's natural one, there will always be pesky situations like yours.

Shrinking your redaction rect (height) even more will however do the trick. Your redaction rect can be as small as you want, it doesn't need to cover the word "von". For example, reduce its height by 20% from its top and its bottom:
```python
delta = rect.height * 0,2
page.add_redact_annot(rect + (0,delta,0,-delta))
```

**Reply by tigrankh** [reply]
thanks for getting back.

yeah, i used `fitz.TOOLS.set_small_glyph_heights(True)` for redaction as well.
reducing the height of redaction to the font height of the word does the trick, i was just not sure why redaction ends up being bigger than the drawn rectangle of the same size.

The 20% you mentioned is some exact value that is coming from somewhere, or depending on text and font height the "error of line-spacing" can be different ?

what it means and what happens when "PDF creator was unfortunate enough" ?

my concern is that if i stick to some percentage or the font height, i could face the same problem when the line has multiple font heights. well, i guess the solution to that could be to use the smallest font size in the line.

**Reply by JorjMcKie** [reply]
>  unfortunate enough:

Example font Helvetica: like all fonts, it has the so-called "ascender" and "descender" properties, which (roughly) indicate how large is the portion above and below the baseline of each character, see this picture from the documentation: https://pymupdf.readthedocs.io/en/latest/_images/img-asc-desc.png
In this case we have:
```python
In [1]: import fitz
In [2]: helv=fitz.Font("helv")
In [3]: helv.ascender,helv.descender
Out[3]: (1.0750000476837158, -0.29899999499320984)
In [4]: # natural line height = fontsize * (ascender - descender)
In [5]: 10 * (helv.ascender - helv.descender)
Out[5]: 13.740000426769257
In [6]:
```
Means, that for fontsize 10, the "natural" vertical distance between two lines should be 13.74.
But when outputting text, you can position stuff where ever you want and e.g. choose to ignore your font's properties.
Some guys even simulate bold text with the regular font, by writing the same text twice but with a tiny offset. This saves a few KB of PDF file size, because you don't pull in another font version for boldness.
But imagine you want to redact this!

> well, i guess the solution to that could be to use the smallest font size in the line.

As I wrote: you can choose a redact rect as small as you want, as long as it is not empty. It only counts, that the to-be-erased characters sufficiently overlap it.
My 20% was just pulled out of the air.

But I am not sure we have no misconception among us:
You must set small glyph heights at the start of everything else - especially before you start searching for text or extracting text. They all are influenced by this global parameter setting.

**Reply by tigrankh** [reply]
Thanks for the detailed explanation.
sure, i understand that size of the redaction rectangle isn't technically that important. but for my use case redaction size should be comparable to the word size.

the small_glyph_heights, i enable right before loading the binary into fitz Document.

**Reply by JorjMcKie** [reply]
> the small_glyph_heights, i enable right before loading the binary into fitz Document.

If you do that, every word is encircled like this:
![grafik](https://user-images.githubusercontent.com/8290722/178725378-a222adf5-38f2-4877-a0e2-e57b301c5258.png)

instead of like this:
![grafik](https://user-images.githubusercontent.com/8290722/178725247-b3809b41-3bb7-4b56-86ed-8197d39ae534.png)

You must have a very unfortunate situation in your doc if you still see inter-line overlaps.

**Reply by tigrankh** [reply]
I guess i indeed found a bad scenario.
When i use draw_rect() with the new values affected by small_glyph_heights, i see that the rectangles don't overlap, just like in your red rectangles. but as soon as i apply redaction on the word it erases the word.

in my experiment i had to use the font size to avoid overlaps

**Reply by JorjMcKie** [reply]
If you used a red rectangle as the redaction rectangle and still got unwanted erasures, your PDF must have really extremely compacted text.
If you want, send me an example page - I still wonder was is going on.
Also don't understand your mention of font size ...

**Reply by tigrankh** [reply]
so the height of rectangle computed by small_glyph_heights was 7.92, which was still causing erasing of the word below.
I used `page.get_fonts()` and saw that there are 4 fonts used throughout the page and one of them has fontsize of 7. so i assumed that its this same word.
I have set the height of redaction rectangle to 7 and it wasn't erasing the word below it anymore.

I'm sorry, but i cant share the PDF.
If you have ideas in mind, i'm happy to do whatever experiments needed to debug this.

Thank you very much for help !

**Reply by JorjMcKie** [reply]
> eight of redaction rectangle to 7

this is a variation of my 20% suggestion further up. If that works, fine.
